### ISSUE TITLE
Safari "Block all cookies" setting breaks JavaScript scripts

### ISSUE THREAD

**Issue by krzksz** [magento/magento2#13865]
### Preconditions

1. Magento 2.2.2 but potentially all versions are affected
2. _Block all cookies_ setting enabled in Safari privacy options.

### Steps to reproduce

1. Open Safari browser and go to _Preferences_
2. In _Privacy_ tab enable _Block all cookies_ setting.
3. Open any Magento shop.

### Expected result

1. At least the popup with "cookies disabled" message is being displayed.

### Actual result

1. There are multiple `SecurityError (DOM Exception 18): The operation is insecure.` errors reported in browser's console, some of the scripts work but I can't add anything to the cart.

So it seems like having above setting enabled in Safari will result in above error every time any script is trying to access either `window.localStorage` or `window.sessionStorage` including reassigning it. As a result following lines throw errors and most probably prevent shop from functioning properly:
https://github.com/magento/magento2/blob/c06da29dbf8cbbab5bb40ad690563b5554e78c60/app/code/Magento/Theme/view/frontend/templates/js/polyfill.phtml#L128
https://github.com/magento/magento2/blob/9c14af7a97bcf338662de365d2613c7ca9dbf8c6/app/code/Magento/Ui/view/base/web/js/lib/core/storage/local.js#L14

and also this part of https://github.com/magento/magento2/blob/2.2-develop/lib/web/jquery/jquery.storageapi.min.js: 
```javascript
    function f(e) {
        if (!window[e]) // Error thrown here
            return !1;
        var t = "jsapi";
        try {
            return window[e].setItem(t, t), window[e].removeItem(t), !0
        } catch (r) {
            return !1
        }
    }
```

I am not sure what would be your preferred solution to this problem but I can gladly prepare a PR when we agree on how we should solve it 👍 .

**Comment by VladimirZaets** [comment]
Hi @krzksz, Magento can't correctly work in the case when cookies are disabled, but I agree that we should give some notification to the user about it. I think the solution with notification in modal window is ok. We should add JS script that will check cookie state when JS application is bootstrapping, for this we can use "cookieEnabled" property in navigation object.

**Comment by ishakhsuvarov** [comment]
Hi @krzksz 
Are you still interested in preparing a PR for this case?

**Comment by raulvOnestic91** [comment]
@magento i am working on it

**Comment by slavvka** [comment]
Hi @krzksz. Thank you for your report. 
The issue has been fixed in magento/magento2#25324 by @raulvOnestic91 in 2.4-develop branch
Related commit(s):
 - [b6bd937a8276924191c8bf753889941636cecd2d](https://github.com/magento/magento2/commit/b6bd937a8276924191c8bf753889941636cecd2d)
 - [7e527d8706a4590075219fb10080e009d8da3422](https://github.com/magento/magento2/commit/7e527d8706a4590075219fb10080e009d8da3422)
 - [60117ee2523187a546cf0aee5b1c4d1ef8ee49b8](https://github.com/magento/magento2/commit/60117ee2523187a546cf0aee5b1c4d1ef8ee49b8)
 - [0c7ac0c7fad2963b18a428af6387f88b3f94568d](https://github.com/magento/magento2/commit/0c7ac0c7fad2963b18a428af6387f88b3f94568d)
 - [7316201c309abdb8af2b4c68f851289e678d6a1d](https://github.com/magento/magento2/commit/7316201c309abdb8af2b4c68f851289e678d6a1d)
 - [ab7c9dccb6e608992e036d45d633a8f25d731a79](https://github.com/magento/magento2/commit/ab7c9dccb6e608992e036d45d633a8f25d731a79)
 - [0ed2bcdf447d1edd60326670a4327d3be66b72d5](https://github.com/magento/magento2/commit/0ed2bcdf447d1edd60326670a4327d3be66b72d5)
 - [1692711ef4ab9368c58b3d7a640ebf995818c53d](https://github.com/magento/magento2/commit/1692711ef4ab9368c58b3d7a640ebf995818c53d)
 - [78e138675193a41a65ee2f7f8b96fae21101a016](https://github.com/magento/magento2/commit/78e138675193a41a65ee2f7f8b96fae21101a016)
 - [1b1c8bf17bfd4b1a252395f33593faedbcce49f1](https://github.com/magento/magento2/commit/1b1c8bf17bfd4b1a252395f33593faedbcce49f1)
 - [14112c0c79921773973a088731a7cce8a1d598da](https://github.com/magento/magento2/commit/14112c0c79921773973a088731a7cce8a1d598da)
 - [f7d91932bb77b759eb448dd352c722e9febdafb5](https://github.com/magento/magento2/commit/f7d91932bb77b759eb448dd352c722e9febdafb5)

The fix will be available with the upcoming 2.4.0 release.
